### ISSUE TITLE
Not working with markdown code chunk with language name

### ISSUE THREAD

**Issue by renkun-ken** [jimhester/lintr#472]
lintr does not seem to work with Rmd file with markdown code chunk with language name specified. Following are examples that show the problem.

* doc1.Rmd

````
# hello

Hello, world!

```{r}
x<-rnorm(100)
y<-rnorm(100)
m <- lm(y~x)
```
````

```
> lintr::lint("doc1.Rmd")                                                       
/Users/ken/Workspaces/test/vscode-r/doc1.Rmd:6:2: style: Put spaces around all infix operators.
x<-rnorm(100)
~^~~
/Users/ken/Workspaces/test/vscode-r/doc1.Rmd:7:2: style: Put spaces around all infix operators.
y<-rnorm(100)
~^~~
```

* doc2.Rmd

````
# hello

Hello, world!

```{r}
x<-rnorm(100)
y<-rnorm(100)
m <- lm(y~x)
```

```
x <- rnorm(100)
y <- rnorm(100)
m <- lm(y ~ x)
```
````

```
> lintr::lint("doc2.Rmd")                                                       
/Users/ken/Workspaces/test/vscode-r/doc2.Rmd:6:2: style: Put spaces around all infix operators.
x<-rnorm(100)
~^~~
/Users/ken/Workspaces/test/vscode-r/doc2.Rmd:7:2: style: Put spaces around all infix operators.
y<-rnorm(100)
~^~~
```

* doc3.Rmd

````
# hello

Hello, world!

```{r}
x<-rnorm(100)
y<-rnorm(100)
m <- lm(y~x)
```

```r
x <- rnorm(100)
y <- rnorm(100)
m <- lm(y ~ x)
```
````

```
> lintr::lint("doc3.Rmd")                                                       
Error: Malformed file!
Backtrace:
1: stop("Malformed file!", call. = FALSE)
2: filter_chunk_end_positions(starts = starts, ends = grep(pattern$chunk.end, 
    lines, perl = TRUE))
3: extract_r_source(source_file$filename, source_file$lines)
4: get_source_expressions(filename)
5: lintr::lint("doc1.Rmd")
```

**Comment by russHyde** [comment]
I modified some code related to this a year or so ago. It might be worth having a look in https://github.com/jimhester/lintr/blob/master/R/extract.R - specifically at the regular expression in `defines_knitr_engine` - to see if this can be generalised to handle both ```` ```{r block-name} ```` and ```` ```r block-name ```` forms

**Comment by russHyde** [comment]
Thanks for reporting this.

I just tried fixing this and ended up on a wild-goose chase. It did seem like a strange bug, because lintr identifies R code blocks using knitr's regexes for identifying R code blocks.

If we make a slight change to your `doc3.Rmd`:
~~~~
---
title: "temp"
output: html_document
---

# hello

Hello, world!

```{r}
x<-rnorm(100)
y<-rnorm(100)
m <- lm(y~x)
m
```

```r
x <- rnorm(100)
y <- rnorm(100)
m <- lm(y ~ x)
m
```
~~~~

... and then knit that, the second code chunk does not get passed to the R engine by knitr: Although the contents of the model object `m` are printed out at the end of the first chunk, they are not printed out at the end of the second chunk. knitr treats the second chunk as "code-formatted" text, not as R syntax.

In short, lintr is telling you that the file is malformed, because, according to knitr's rules for identifying R-code blocks, it is malformed: the ` ``` r ` line should have braces in it.

Closing

**Comment by renkun-ken** [comment]
Thanks for investigating in this.

Then it would mean that the Rmd cannot be processed if it includes any text code chunk with specified language that is not intended for running but code demonstration?

For example, the following Rmd should be perfectly valid for `knitr::knit` but it cannot be processed by lintr:

````
---
title: "temp"
output: html_document
---

# hello

Hello, world!

```{r}
x<-rnorm(100)
y<-rnorm(100)
m <- lm(y~x)
m
```

```cpp
void test(int x) {
  return x + 1;
}
```
````

**Comment by russHyde** [comment]
Sorry, I mustn't have understood the original issue. Does the ```` ```cpp ```` line tell knitr to use a formatting engine?
